lib/Reader/readerir.cpp

```diff
@@ -5994,8 +5994,35 @@ void GenIR::switchOpcode(IRNode *Opr) {
   BasicBlock *CurrBlock = LLVMBuilder->GetInsertBlock();
   TerminatorInst *TermInst = CurrBlock->getTerminator();
   SwitchInst *SwitchInstruction = cast<SwitchInst>(TermInst);
-
   SwitchInstruction->setCondition(Opr);
+
+  Type *OprType = Opr->getType();
+  LLVMContext &LLVMContext = *this->JitContext->LLVMContext;
+  Type *Int32Ty = Type::getInt32Ty(LLVMContext);
+  if (OprType != Int32Ty) {
+    // When we made the case values we assumed int32 as the type for the
+    // values. But now we have an operand of a different type. LLVM
+    // requires the values and the switch value to have the same type.
+    // So change the type of the values to be the type of the selector.
+    IntegerType *OprIntType = cast<IntegerType>(OprType);
+
+    // If a small selector type was used, e.g. i8, the type might not
+    // be big enough to hold all the case values. But the value for the
+    // selector is popped off the evaluation stack, so it should be at
+    // least 32 bits. So check that.
+    unsigned OprBitWidth = OprIntType->getBitWidth();
+    assert((OprBitWidth > 32) && "Selector bit width is less then 32");
+
+    // Create new values of the correct type.
+    int NumCases = SwitchInstruction->getNumCases();
+    SwitchInst::CaseIt Case = SwitchInstruction->case_begin();
+    for (size_t I = 0; I < NumCases; I++) {
+      ConstantInt *Value = ConstantInt::get(
+          *JitContext->LLVMContext, APInt(OprBitWidth, (uint64_t)I, false));
```

Comment:
Now that you mention it, the switch printed out when the verification failed looks odd since it does not have a value for 0. I've gone back into the debugger and verified that the switch at the 
point where I modify the types is:

```
  switch i64 %23, label %24 [
    i32 0, label %24
    i32 1, label %25
    i32 2, label %27
    i32 3, label %29
    i32 4, label %30
    i32 5, label %31
    i32 6, label %36
  ]
```

so the switch gets modified later. However seeing this I do think the code should be written
to be more robust. It should have as a post-condition that the case values are preserved 
but that all their types are modified to match. I will make those changes.